File: common/src/test/java/com/hartwig/actin/molecular/sort/driver/DriverComparatorTest.java
package com.hartwig.actin.molecular.sort.driver;

import static org.junit.Assert.assertEquals;

import java.util.List;

import com.google.common.collect.Lists;
import com.hartwig.actin.molecular.datamodel.driver.Driver;
import com.hartwig.actin.molecular.datamodel.driver.DriverLikelihood;
import com.hartwig.actin.molecular.datamodel.evidence.ActionableEvidence;
import com.hartwig.actin.molecular.datamodel.evidence.TestActionableEvidenceFactory;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.junit.Test;

public class DriverComparatorTest {

    @Test
    public void canSortDrivers() {
        Driver driver1 = create(true, "event 1", DriverLikelihood.HIGH, TestActionableEvidenceFactory.createEmpty());
        Driver driver2 = create(true, "event 1", DriverLikelihood.MEDIUM, TestActionableEvidenceFactory.createEmpty());
        Driver driver3 = create(true, "event 2", DriverLikelihood.MEDIUM, TestActionableEvidenceFactory.createExhaustive());
        Driver driver4 = create(true, "event 2", DriverLikelihood.MEDIUM, TestActionableEvidenceFactory.createEmpty());
        Driver driver5 = create(false, "event 1", DriverLikelihood.HIGH, TestActionableEvidenceFactory.createEmpty());

        List<Driver> drivers = Lists.newArrayList(driver4, driver5, driver1, driver2, driver3);
        drivers.sort(new DriverComparator());

        assertEquals(driver1, drivers.get(0));
        assertEquals(driver2, drivers.get(1));
        assertEquals(driver3, drivers.get(2));
        assertEquals(driver4, drivers.get(3));
        assertEquals(driver5, drivers.get(4));
    }

    @NotNull
    private static Driver create(boolean isReportable, @NotNull String event, @Nullable DriverLikelihood driverLikelihood,
            @NotNull ActionableEvidence evidence) {
        return new Driver() {
            @Override
            public boolean isReportable() {
                return isReportable;
            }

            @NotNull
            @Override
            public String event() {
                return event;
            }

            @Nullable
            @Override
            public DriverLikelihood driverLikelihood() {
                return driverLikelihood;
            }

            @NotNull
            @Override
            public ActionableEvidence evidence() {
                return evidence;
            }


            @Override
            public String toString() {
                return isReportable + " " + event + " " + driverLikelihood;
            }
        };
    }
}

File: common/src/test/java/com/hartwig/actin/molecular/sort/driver/DriverComparatorTest.kt
package com.hartwig.actin.molecular.sort.driver

import com.hartwig.actin.molecular.datamodel.driver.Driver
import com.hartwig.actin.molecular.datamodel.driver.DriverLikelihood
import com.hartwig.actin.molecular.datamodel.evidence.ActionableEvidence
import com.hartwig.actin.molecular.datamodel.evidence.TestActionableEvidenceFactory.createEmpty
import com.hartwig.actin.molecular.datamodel.evidence.TestActionableEvidenceFactory.createExhaustive
import org.assertj.core.api.Assertions.assertThat
import org.junit.Test

class DriverComparatorTest {

    @Test
    fun `Should sort drivers`() {
        val driver1 = create(true, "event 1", DriverLikelihood.HIGH, createEmpty())
        val driver2 = create(true, "event 1", DriverLikelihood.MEDIUM, createEmpty())
        val driver3 = create(true, "event 2", DriverLikelihood.MEDIUM, createExhaustive())
        val driver4 = create(true, "event 2", DriverLikelihood.MEDIUM, createEmpty())
        val driver5 = create(false, "event 1", DriverLikelihood.HIGH, createEmpty())

        val drivers = listOf(driver4, driver5, driver1, driver2, driver3).sortedWith(DriverComparator())

        assertThat(drivers[0]).isEqualTo(driver1)
        assertThat(drivers[1]).isEqualTo(driver2)
        assertThat(drivers[2]).isEqualTo(driver3)
        assertThat(drivers[3]).isEqualTo(driver4)
        assertThat(drivers[4]).isEqualTo(driver5)
    }

    private fun create(
        isReportable: Boolean, event: String, driverLikelihood: DriverLikelihood?,
        evidence: ActionableEvidence
    ): Driver {
        return object : Driver {
            override fun toString(): String {
                return "$isReportable $event $driverLikelihood"
            }

            override val isReportable: Boolean
                get() = isReportable

            override val event: String
                get() = event

            override val driverLikelihood: DriverLikelihood?
                get() = driverLikelihood

            override val evidence: ActionableEvidence
                get() = evidence
        }
    }
}